File: R/gcp.R
#' Lee un archivo Parquet desde Google Cloud Storage
#'
#' Equivalente a read_parquet_gcp() de Python.
#'
#' @param gcp_path Character. Path en formato "bucket/path/to/file.parquet".
#' @return data.frame
#'
#' @importFrom googleCloudStorageR gcs_get_object
#' @importFrom arrow read_parquet
#' @export
read_parquet_gcp <- function(gcp_path) {
  parts       <- .parse_gcp_path(gcp_path)
  tmp         <- tempfile(fileext = ".parquet")
  on.exit(unlink(tmp))
  googleCloudStorageR::gcs_get_object(
    object_name = parts$prefix,
    bucket      = parts$bucket,
    saveToDisk  = tmp,
    overwrite   = TRUE
  )
  arrow::read_parquet(tmp)
}

#' Escribe un data.frame como Parquet en Google Cloud Storage
#'
#' Equivalente a write_parquet_gcp() de Python.
#'
#' @param df data.frame a guardar.
#' @param gcp_path Character. Path destino en formato "bucket/path/to/file.parquet".
#'
#' @importFrom googleCloudStorageR gcs_upload
#' @importFrom arrow write_parquet
#' @export
write_parquet_gcp <- function(df, gcp_path) {
  parts <- .parse_gcp_path(gcp_path)
  tmp   <- tempfile(fileext = ".parquet")
  on.exit(unlink(tmp))
  arrow::write_parquet(df, tmp, compression = "gzip")
  googleCloudStorageR::gcs_upload(tmp, bucket = parts$bucket, name = parts$prefix)
}

#' Lista archivos en un path de Google Cloud Storage
#'
#' Equivalente a list_files_gcp() de Python.
#'
#' @param gcp_path Character. Path en formato "bucket/prefijo/".
#' @param kwd Character. Filtro opcional por keyword en el nombre del archivo.
#' @return Character vector con los nombres de los blobs.
#'
#' @importFrom googleCloudStorageR gcs_list_objects
#' @export
list_files_gcp <- function(gcp_path, kwd = NULL) {
  parts <- .parse_gcp_path(gcp_path)
  files <- googleCloudStorageR::gcs_list_objects(
    bucket = parts$bucket,
    prefix = parts$prefix
  )
  names <- files$name
  if (!is.null(kwd)) names <- names[grepl(kwd, names, fixed = TRUE)]
  names
}

# Helper interno: separa bucket y prefix de un gcp_path
.parse_gcp_path <- function(gcp_path) {
  parts  <- strsplit(gcp_path, "/", fixed = TRUE)[[1]]
  bucket <- parts[1]
  prefix <- paste(parts[-1], collapse = "/")
  list(bucket = bucket, prefix = prefix)
}

File: python/apoyo/gcp.py
from io import BytesIO

import pandas as pd
from google.cloud import storage


def _parse_gcp_path(gcp_path: str) -> tuple[str, str]:
    """Separa un path GCS en (bucket_name, prefix)."""
    parts = gcp_path.split("/", 1)
    bucket_name = parts[0]
    prefix = parts[1] if len(parts) > 1 else ""
    return bucket_name, prefix


def read_parquet_gcp(gcp_path: str) -> pd.DataFrame:
    """
    Lee un archivo Parquet desde GCS.

    Parameters
    ----------
    gcp_path : str
        Path en formato 'bucket/path/to/file.parquet'.

    Returns
    -------
    pd.DataFrame
    """
    bucket_name, prefix = _parse_gcp_path(gcp_path)
    cli = storage.Client()
    data = cli.bucket(bucket_name).get_blob(prefix).download_as_bytes()
    return pd.read_parquet(BytesIO(data))


def write_parquet_gcp(df: pd.DataFrame, gcp_path: str) -> None:
    """
    Escribe un DataFrame como Parquet en GCS.

    Parameters
    ----------
    df : pd.DataFrame
    gcp_path : str
        Path destino en formato 'bucket/path/to/file.parquet'.
    """
    bucket_name, prefix = _parse_gcp_path(gcp_path)
    buf = BytesIO()
    df.to_parquet(buf, engine="pyarrow")
    buf.seek(0)
    storage.Client().bucket(bucket_name).blob(prefix).upload_from_file(buf)


def list_files_gcp(gcp_path: str, kwd: str = None) -> list[str]:
    """
    Lista archivos en un path de GCS, con filtro opcional por keyword.

    Parameters
    ----------
    gcp_path : str
        Path en formato 'bucket/prefijo/'.
    kwd : str, optional
        Si se indica, retorna solo los archivos cuyo nombre contiene kwd.

    Returns
    -------
    list[str]
        Lista de nombres de blob.
    """
    bucket_name, prefix = _parse_gcp_path(gcp_path)
    blobs = storage.Client().bucket(bucket_name).list_blobs(prefix=prefix)
    names = [b.name for b in blobs]
    if kwd:
        names = [n for n in names if kwd in n]
    return names
